### ISSUE TITLE
Reproducing results on CUB-200-2011

### ISSUE THREAD

**Issue by Sebbe-Stouten** [PeiqinZhuang/API-Net#7]
Hello, I have been trying to reproduce the benchmark results reported in your paper, more specifically the ones where you train a ResNet-101 backbone with an API-net on top on the CUB-200-2011 dataset.

My first attempt was on a 16G GPU, so I could only set --n_classes 9, and --n_samples 3. I reached 87.0% validation accuracy, whereas in the paper a 88.6% was reported. The obvious explanation for this was that reducing the number of classes and samples, reduced API-net's performance, which makes sense.
So I have tried some experiments on a larger GPU, so that I could use more classes and samples, but this didn't give me the expected improvement. 

When I set --n_classes 25, --n_samples 3, I only got 85.9% validation accuracy. (Notice that this is even lower than my first 9x3 experiment!) I even tried using Adam instead of SGD, but no improvement there. 
So, I tried --n_classes 19, --n_samples 4, but got similar results.
I noticed that the training curves flatten early: already around epoch 30 training accuracy is 100%, and the loss and validation accuracy hardly change anymore.

Any ideas about what could go wrong or how to improve my results would be very welcome, thanks!

**Comment by PeiqinZhuang** [comment]
Hi. Frankly speaking, I have no ideas about the failures of your experiments. Could you please detail the changes you have made and the training process, i.e. how many gpus you used. And if possible, you may need to check your data once again.

**Comment by Sebbe-Stouten** [comment]
Hi, thanks for your answer. As far as I can see, I have made 3 changes:

1. Naively changed the source code from Python 2.7 to Python 3.6 by replacing all print statements. I am not sure if this could cause major changes in the training process, but I didn't get any warnings about something like that. Maybe this is important though, I don't know if there is a particular reason you used Python 2.7 and Pytorch 0.4?
2. Instead of building the validation dataloader after each training epoch, I just build it once before the start of the training
3. Made sure that the BalancedBatchSampler uses its last sample (and thus validation is triggered) by setting the `<` to `<=` in https://github.com/PeiqinZhuang/API-Net/blob/c4996c7fec3fbd46fe2873f74e8ba154218b7e7f/datasets.py#L85 

I ran all experiments on 1 GPU. The small batch size on one with 16G memory, the larger batch sizes on a 40G one.

I had a look at the data used, and there seems to be no issue there: the CUB dataset contains all the images it should, train-val split is correct, and in `read_error.txt` I only got 2 reading errors, so that is unlikely to be the issue as well.
